### PR TITLE
Percents + visualization

### PR DESCRIPTION
**Pull request by haaavk**
I managed to do visualization similar to what is in Plausible.
Shynet - percents new - Screenshot 2021-05-16 21:11:03

I'm not very happy about style attributes. I could use some help with that.

### CODE REVIEW

**Comment by CasperVerswijvelt** [comment]
I would put this style on the 'bar' element to make it look a little spacier/better:
```
    top: 4px;
    left: 4px;
    height: calc(100% - 8px);
```

Maybe the width should also be based on the maximum amount of sessions for an element:
so in your screenshot, 100% would be 44 in the browsers category

And then maybe add those bars for country and operating system tables

**Comment by CasperVerswijvelt** [comment]
Seems like I checked out an outdated version where not all tables had the bars yet, my bad

**Comment by CasperVerswijvelt** [comment]
I do think a width relative to the element with most sessions would be nice, so the differences in bar lenghts are more clear at first glance

File: shynet/dashboard/templates/dashboard/pages/service.html [hunk 1]
```diff
@@ -109,8 +109,24 @@
...
                 <tr>
-                    <td class="truncate w-full max-w-0">{{location.location|default:"Unknown"|urldisplay}}</td>
-                    <td class="rf">{{location.count|intcomma}}</td>
+                    <td class="truncate w-full max-w-0 relative">
+                        <div
+                            class="absolute h-6"
+                            style="width: {{location.count|percent:stats.hit_count}}; top: 8px; left: 8px; background-color: var(--color-urge-200-fallback)"
+                        >
+                        </div>
+                        <div class="absolute" style="top: 8px; left: 8px">
```
Comment (CasperVerswijvelt) [inline shynet/dashboard/templates/dashboard/pages/service.html:118]: replace `class="absolute" style="top: 8px; left: 8px"` with `class="relative"` for same effect

Same thing for the other tables

**Comment by CasperVerswijvelt** [comment]
I made my suggested changes in my repo on the [percent-visualization-changes](https://github.com/CasperVerswijvelt/shynet/tree/percent-visualization-changes) branch if you want to use them.

About the relative maximum: Plausible also visualizes the top element as a bar with width 100%, what do you think @milesmcc ?

**Comment by milesmcc** [comment]
Yeah I do think it would be good to scale the bars so that the first/highest element has full width. Makes it easier to distinguish the relative sizes.

**Comment by haaavk** [comment]
I think showing actual percent feels more natural. It's easier to understand and shows more information.
With absolute max it is possible to compare date between cards. 
Relative max make it easier to compare date in one card and it's notable advantage.
On the other hand it may confuse some users and makes it impossible to compare date from different cards. 
In my opinion there are more arguments for absolute max.

File: shynet/dashboard/templates/dashboard/pages/service.html [hunk 2]
```diff
@@ -201,9 +261,24 @@
...
-                    <td class="flex items-center truncate w-full max-w-0" title="{{browser.browser|default:'Unknown'}}">
-                        {{browser.browser|iconify}}<span>{{browser.browser|default:"Unknown"}}</span></td>
-                    <td class="rf">{{browser.count|intcomma}}</td>
+                    <td class="flex items-center truncate w-full max-w-0 relative" title="{{browser.browser|default:'Unknown'}}">
+                        <div
+                            class="absolute h-6"
+                            style="width: {{browser.count|percent:stats.session_count}}; top: 6px; left: 6px; height: calc(100% - 12px); background-color: var(--color-urge-200-fallback)"
+                        >
+                        </div>
+                        <div class="relative">
```
Comment (CasperVerswijvelt) [inline shynet/dashboard/templates/dashboard/pages/service.html:270]: add back `flex items-center` classes to correctly align icon and text better

File: shynet/dashboard/templates/dashboard/pages/service.html [hunk 3]
```diff
@@ -177,10 +223,24 @@
...
                 <tr>
-                    <td class="flex items-center truncate w-full max-w-0" title="{{os.os|default:'Unknown'}}">
-                        {{os.os|iconify}}<span>{{os.os|default:"Unknown"}}</span>
+                    <td class="flex items-center truncate w-full max-w-0 relative" title="{{os.os|default:'Unknown'}}">
+                        <div
+                            class="absolute h-6"
+                            style="width: {{os.count|percent:stats.session_count}}; top: 6px; left: 6px; height: calc(100% - 12px); background-color: var(--color-urge-200-fallback)"
+                        >
+                        </div>
+                        <div class="relative">
```
Comment (CasperVerswijvelt) [inline shynet/dashboard/templates/dashboard/pages/service.html:232]: add back `flex items-center` classes to correctly align icon and text better

File: shynet/dashboard/templates/dashboard/pages/service.html [hunk 4]
```diff
@@ -109,8 +109,24 @@
             <tbody>
                 {% for location in stats.locations %}
                 <tr>
-                    <td class="truncate w-full max-w-0">{{location.location|default:"Unknown"|urldisplay}}</td>
-                    <td class="rf">{{location.count|intcomma}}</td>
+                    <td class="truncate w-full max-w-0 relative">
+                        <div
+                            class="absolute h-6"
+                            style="width: {{location.count|percent:stats.hit_count}}; top: 6px; left: 6px; height: calc(100% - 12px); background-color: var(--color-urge-200-fallback)"
```
Comment (CasperVerswijvelt) [inline shynet/dashboard/templates/dashboard/pages/service.html:115]: Maybe move style attributes to css file to minimize duplicate code like in [this commit](https://github.com/CasperVerswijvelt/shynet/commit/19520279de8e5f89794bb23e14dd9596ae9c3868)

Comment (CasperVerswijvelt) [inline shynet/dashboard/templates/dashboard/pages/service.html:115]: `left: 0;` to align bar with the table header/body seperator line and make it not get cut off by the `td` tag when bar is full or close to full width

**Review by CasperVerswijvelt — CHANGES_REQUESTED** [review]

**Comment by CasperVerswijvelt** [comment]
I do think, for example in this case, that making the bars as big as possible would improve the clarity of the differences between bar widths a lot. 

I understand that you would then no longer be able to compare the bars between tables, but that is not something I do anyway. Can't speak for others though. 

I just think there is a lot of _unused_ whitespace that could be used to make the differences between the bars clearer.

image

**Comment by haaavk** [comment]
We just want to achieve different things with this visualization. Maybe it should be customizable in settings?

**Comment by CasperVerswijvelt** [comment]
> We just want to achieve different things with this visualization. Maybe it should be customizable in settings?

That sounds good to me :)

File: shynet/shynet/settings.py [hunk 5]
```diff
@@ -346,7 +349,12 @@
...
-LOCATION_URL = os.getenv("LOCATION_URL", "https://www.openstreetmap.org/?mlat=$LATITUDE&mlon=$LONGITUDE")
+LOCATION_URL = os.getenv(
+    "LOCATION_URL", "https://www.openstreetmap.org/?mlat=$LATITUDE&mlon=$LONGITUDE"
+)
 
 # How many services should be displayed on dashboard page?
 DASHBOARD_PAGE_SIZE = int(os.getenv("DASHBOARD_PAGE_SIZE", "5"))
+
+# Should background bars be scaled to full width?
+USE_RELATIVE_MAX_IN_BAR_VISUALIZATION = True
```
Comment (CasperVerswijvelt) [inline shynet/shynet/settings.py:360]: `USE_RELATIVE_MAX_IN_BAR_VISUALIZATION = os.getenv("USE_RELATIVE_MAX_IN_BAR_VISUALIZATION", "True") == "True"`

File: shynet/dashboard/templates/dashboard/includes/bar.html [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+        style="width: {{count|percent:max}}; top: 6px; left: 0px; height: calc(100% - 12px); background-color: var(--color-urge-200-fallback)"
+    >
+    </div>
+{% else %}
+    <div
+        class="absolute h-6"
+        style="width: {{count|percent:total}}; top: 6px; left: 0px; height: calc(100% - 12px); background-color: var(--color-urge-200-fallback)"
+    >
+    </div>
+{% endif %}
```
Comment (CasperVerswijvelt) [inline shynet/dashboard/templates/dashboard/includes/bar.html:15]: simplify to 
```
<div
    class="absolute h-6"
    style="width: {% if use_relative_max %}{{count|percent:max}}{% else %}{{count|percent:total}}{% endif %}; top: 6px; left: 0px; height: calc(100% - 12px); background-color: var(--color-urge-200-fallback)"
></div>
```
?

File: shynet/dashboard/templates/dashboard/pages/service.html [hunk 7]
```diff
@@ -186,12 +178,8 @@
...
                     <td class="truncate w-full max-w-0 relative" title="{{country.country|country_name}}">
-                        <div
-                            class="absolute h-6"
-                            style="width: {{country.count|percent:stats.session_count}}; top: 6px; left: 6px; height: calc(100% - 12px); background-color: var(--color-urge-200-fallback)"
-                        >
-                        </div>
-                        <div class="relative">
+                        {% include 'dashboard/includes/bar.html' with count=country.count max=stats.countries.0.count total=stats.session_count %}
+                        <div class="relative flex items-center">
                             <span class="{{country.country|flag_class}}"></span> {{country.country|country_name}}
```
Comment (CasperVerswijvelt) [inline shynet/dashboard/templates/dashboard/pages/service.html:183]: Add `flex-none` class here to make flag not get squished

File: shynet/dashboard/templates/dashboard/includes/bar.html (hunk 6 above)
Comment (haaavk) [reply shynet/dashboard/templates/dashboard/includes/bar.html:15]: I know it probably should be done this way. This line just looks so ugly.

Comment (CasperVerswijvelt) [reply shynet/dashboard/templates/dashboard/includes/bar.html:15]: I know :p
It's ugly code vs duplicate code

File: shynet/dashboard/templates/dashboard/pages/service.html (hunk 7 above)
Comment (CasperVerswijvelt) [reply shynet/dashboard/templates/dashboard/pages/service.html:183]: `flex-none` needs to be on the 
```
<span class="{{country.country|flag_class}}"></span> {{country.country|country_name}}
```
line, not the one above

**Comment by haaavk** [comment]
I think it's ready to be merged.

**Comment by milesmcc** [comment]
This is fantastic. Merged. Will be in the next release! :)

**Pull request merged**